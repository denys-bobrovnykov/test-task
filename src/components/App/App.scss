body {
  color: #333;
  font-family: Georgia, 'Times New Roman', Times, serif;
}
main {
  display: flex;
  min-width: 600px;
}
.container {
  box-shadow: inset;
  padding: 20px;
}
a {
  position: absolute;
  top: 50%;
  transform: translateY(-50%);
  width: 100%;
  h1 {
    text-align: center;
  }
}
