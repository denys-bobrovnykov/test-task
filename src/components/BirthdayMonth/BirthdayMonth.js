import { useSelector } from 'react-redux';
import { getMonth } from '../../service/date-utils';
import { SelectedItem } from '../SelectedItem/SelectedItem';
import './BirthdayMonth.scss';

export const BirthdayMonth = (props) => {
  const state = useSelector((state) => state.list);

  return (
    <div className="BirthdayMonth">
      <h3>{props.month}</h3>
      <ul>
        {state.selected
          .filter((el) => getMonth(el.dob) === props.month)
          .map((el) => (
            <SelectedItem key={el.id} item={el} />
          ))}
      </ul>
    </div>
  );
};
