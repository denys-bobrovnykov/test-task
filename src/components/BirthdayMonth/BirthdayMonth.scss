.BirthdayMonth {
  ul {
    padding: 0;
    li {
      margin-left: 20px;
    }
  }
}
