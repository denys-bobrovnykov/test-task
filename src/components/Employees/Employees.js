import { Row } from '../Row/Row';
import './Employees.scss';

export const Employees = () => {
  // Create alphabet
  const letters = 'abcdefghijklmnopqrstuvwxyz'.toUpperCase().split('');
  const renderRows = () => {
    return letters.map((letter) => <Row key={letter} letter={letter} />);
  };

  return (
    <div className="container Employees">
      <h2>Employees</h2>
      <div className="scroller">{renderRows()}</div>
    </div>
  );
};
