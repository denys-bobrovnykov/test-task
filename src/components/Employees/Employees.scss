.Employees {
  align-content: flex-start;
  display: flex;
  flex-wrap: wrap;
  width: 70%;
  height: 100%;
  h2 {
    text-align: center;
    width: 100%;
  }
  .scroller {
    display: flex;
    width: 100%;
    height: 100%;
    overflow: scroll;
  }
  ul {
    width: 100%;
  }
}
