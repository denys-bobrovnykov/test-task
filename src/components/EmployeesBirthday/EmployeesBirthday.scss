.EmployeesBirthday {
  align-content: flex-start;
  justify-content: center;
  display: flex;
  flex-wrap: wrap;
  width: 40%;
  h2 {
    text-align: center;
  }
  .wrapper {
    border-left: 1px solid black;
    height: 100%;
    padding-left: 20px;
  }
  .presentation {
    border-top: 1px solid black;
  }
  ul {
    list-style-type: disc;
  }
}
