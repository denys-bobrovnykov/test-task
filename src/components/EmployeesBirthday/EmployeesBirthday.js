import { useSelector } from 'react-redux';
import { getMonth } from '../../service/date-utils';
import { BirthdayMonth } from '../BirthdayMonth/BirthdayMonth';
import './EmployeesBirthday.scss';

export const EmployeesBirthday = () => {
  const state = useSelector((state) => state.list);
  const months = [];

  const renderMonths = () => {
    return months.length > 0 ? (
      months.map((el) => <BirthdayMonth key={el} month={el} />)
    ) : (
      <h3 className="empty-title">No selected employees</h3>
    );
  };
  // Select all months which are in selected state
  state.selected.forEach((el) => {
    const month = getMonth(el.dob);
    if (months.indexOf(month) === -1) {
      months.push(month);
    }
  });

  return (
    <div className="container EmployeesBirthday">
      <h2>Employees birthday</h2>
      <div className="wrapper">
        <div className="presentation">{renderMonths()}</div>
      </div>
    </div>
  );
};
