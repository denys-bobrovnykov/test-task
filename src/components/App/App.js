import { useEffect } from 'react';
import { useDispatch } from 'react-redux';
import { BrowserRouter as Router, Route, Switch, Link } from 'react-router-dom';
import { fetchData } from '../../redux/actions/actionCreators';
import { Employees } from '../Employees/Employees';
import { EmployeesBirthday } from '../EmployeesBirthday/EmployeesBirthday';
import './App.scss';

function App() {
  const dispatch = useDispatch();
  useEffect(() => {
    dispatch(fetchData());
  }, []);
  return (
    <Router>
      <Switch>
        <Route path="/" exact>
          <Link id="employees-link" to="/employees">
            <h1>Go To Employees</h1>
          </Link>
        </Route>
        <Route path="/employees" exact>
          <main className="App">
            <Employees />
            <EmployeesBirthday />
          </main>
        </Route>
      </Switch>
    </Router>
  );
}

export default App;
