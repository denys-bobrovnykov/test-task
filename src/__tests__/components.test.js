import { render, cleanup } from '@testing-library/react';
import { act, create } from 'react-test-renderer';
import React from 'react';
import configureStore from 'redux-mock-store';
import { Provider } from 'react-redux';
import thunk from 'redux-thunk';
import { testData } from '../test-data/testData';
import * as states from '../test-data/states';
import App from '../components/App/App';
import { Item } from '../components/Item/Item';
import { Row } from '../components/Row/Row';
import { Employees } from '../components/Employees/Employees';
import { EmployeesBirthday } from '../components/EmployeesBirthday/EmployeesBirthday';
import { BirthdayMonth } from '../components/BirthdayMonth/BirthdayMonth';
import { SelectedItem } from '../components/SelectedItem/SelectedItem';

const mockStore = configureStore([thunk]);

describe('App tests', () => {
  let store;
  beforeEach(() => {
    store = mockStore({ list: states.initialState });
  });
  afterEach(() => {
    store = null;
    cleanup();
  });

  test('App rendering home page', () => {
    const snapshot = create(
      <Provider store={store}>
        <App />
      </Provider>
    );
    expect(snapshot.toJSON()).toMatchSnapshot();
  });
});

describe('Employee list item', () => {
  let store;
  beforeEach(() => {
    store = mockStore({ list: states.initialState });
  });
  afterEach(() => (store = null));

  test('Renders initial state unchecked', () => {
    render(
      <Provider store={store}>
        <Item item={testData[0]} />
      </Provider>
    );
    expect(document.querySelector('input')).not.toBeChecked();
  });

  test('Renders selected state checked', () => {
    store = mockStore({ list: states.selectedState });
    const snapshot = render(
      <Provider store={store}>
        <Item item={testData[0]} />
      </Provider>
    );
    expect(document.querySelector('input')).toBeChecked();
  });
});

describe('Employees', () => {
  let store;
  beforeEach(() => {
    store = mockStore({ list: states.initialState });
  });
  afterEach(() => (store = null));

  test('Renders alphabet as <Row props=Letter /> elements', () => {
    let component;
    act(() => {
      component = create(
        <Provider store={store}>
          <Employees />
        </Provider>
      );
    });
    const alphabet = 'abcdefghijklmnopqrstuvwxyz'.toUpperCase().split('');
    const rows = component.root.findAllByType(Row);
    rows.forEach((el, index) => {
      expect(el.props.letter).toEqual(alphabet[index]);
    });
    component.unmount();
  });
});

describe('Employees Birthday', () => {
  let store;
  let component;
  beforeEach(() => {
    store = mockStore({ list: states.initialState });
  });
  afterEach(() => {
    store = null;
    component = null;
  });

  test('Renders "No selected employees" with nothing selected', () => {
    act(() => {
      component = create(
        <Provider store={store}>
          <EmployeesBirthday />
        </Provider>
      );
    });
    expect(component.root.findByType('h3').props.children).toContain('No selected employees');
    component.unmount();
  });

  test('Renders months as <BirthdayMonth /> with corresponding month text when selected', () => {
    store = mockStore({ list: states.selectedState });
    act(() => {
      component = create(
        <Provider store={store}>
          <EmployeesBirthday />
        </Provider>
      );
    });
    const dob = store.getState().list.selected[0].dob;
    const month = new Date(dob).toLocaleString('default', { month: 'long' });
    expect(component.root.findAllByType(BirthdayMonth)[0].props.month).toEqual(month);
    component.unmount();
  });

  test('Renders two people with same month bithday in the same <BirthdayMonth />', () => {
    store = mockStore({ list: states.selectedFebruary });
    act(() => {
      component = create(
        <Provider store={store}>
          <EmployeesBirthday />
        </Provider>
      );
    });
    const selectedPersons = component.root.findAllByType(SelectedItem);
    selectedPersons.forEach((person) => {
      expect(new Date(person.props.item.dob).getMonth()).toBe(1);
    });
    component.unmount();
  });
});
