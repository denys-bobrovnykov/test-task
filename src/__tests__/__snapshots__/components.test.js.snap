// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App tests App rendering home page 1`] = `
<a
  href="/employees"
  id="employees-link"
  onClick={[Function]}
>
  <h1>
    Go To Employees
  </h1>
</a>
`;
